app/models/manageiq/providers/openstack/manager_mixin.rb

```diff
@@ -93,25 +98,33 @@ def ems_connect?(password, params, service)
     #   }
     # }
     def verify_credentials(args)
-      endpoint_name = args.dig("endpoints").keys.first
-      endpoint = args.dig("endpoints", endpoint_name)
-      authentication = args.dig("authentications", endpoint_name)
-
-      userid, password = authentication&.values_at('userid', 'password')
-      password = ManageIQ::Password.try_decrypt(password)
-      password ||= find(args["id"]).authentication_password(endpoint_name) if args["id"]
-
-      params = %w[hostname port security_protocol].reduce(
-        [endpoint_name, 'userid'].join('_')   => userid,
-        [endpoint_name, 'password'].join('_') => password
-      ) do |obj, item|
-        obj.merge([endpoint_name, item].join('_') => endpoint.try(:[], item))
+      endpoints = args["endpoints"] || {}
+      authentications = args["authentications"]
+
+      # ceilometer has no additional endpoint info other than what is in the default endpoint
+      # but has to be verified separately
+      if args["event_stream_selection"] == "ceilometer"
+        endpoints["ceilometer"] = endpoints["default"]
+        authentications["ceilometer"] = authentications["default"]
       end
 
-      params.merge!(args.slice('name', 'provider_region', 'api_version', 'uid_ems'))
-      params['event_stream_selection'] = args['event_stream_selection'] if endpoint_name != 'default'
+      endpoints.each do |endpoint_name, endpoint|
+        authentication = authentications[endpoint_name]
+
+        userid, password = authentication&.values_at('userid', 'password')
+        password = ManageIQ::Password.try_decrypt(password)
+        password ||= find(args["id"]).authentication_password(endpoint_name) if args["id"]
 
-      !!raw_connect(password, params.symbolize_keys)
+        params = {
+          "userid"   => userid,
+          "password" => password,
+        }.merge(endpoint.slice("hostname", "port", "security_protocol"))
+
+        params.merge!(args.slice('name', 'provider_region', 'api_version', 'uid_ems'))
```

Comment:
I feel like this would read nicer as...

```suggestion
        endpoint_params = endpoint.slice("hostname", "port", "security_protocol")
        args_params     = args.slice('name', 'provider_region', 'api_version', 'uid_ems')

        params = {
            "userid"   => userid,
            "password" => password
        }.merge(endpoint_params).merge(args_params)
```

or something like that.